src/config/_parser.py

```diff
@@ -99,30 +107,76 @@ def setup_logging(logfile: str):
     logging.info("Configuration file loading...")
 
 
-def setup(filename: str) -> None:
+_hash: str = ''
+
+
+def load_config(filename: str) -> Tuple[Dict, str]:
     """
-    Sets up the module by loading the specified configuration file from disk
+    Loads the configuration file
 
     Args:
-        filename (str): path and filename to load.
+        filename (str): name of configuration file, relative to ./config
+
+    Returns:
+        loaded data and the file's hash.
     """
-    global config
+    path = Path("config") / filename
 
-    path = f"config/{filename}"
+    # create a new hasher
+    hasher = hashlib.sha256()
     # check if the file exists
-    if os.path.exists(path):
-        logging.info(f"Found a file/directory at {filename}'! attempting to load...")
-        with open(path, 'r', encoding="UTF8") as infile:
-            config_dict = toml.load(infile)
-            logging.info("Successfully loaded JSON from file specified!")
-
-        setup_logging(config_dict['logging']['log_file'])
-        config = config_dict
-    else:
-        raise FileNotFoundError(f"Unable to find {filename}")
+    logging.debug(f"Found a file/directory at {path.resolve(strict=True)}'! attempting to load...")
+    with path.open('r', encoding="UTF8") as infile:
+        infile.flush()
+        config_dict = toml.load(infile)
+        logging.info("Successfully loaded from file specified!")
+
+    # update hasher with the raw bytes of the file
+    hasher.update(path.read_bytes())
+
+    del infile
+
+    # digest the file, get its checksum.
+    checksum = hasher.hexdigest()
+
+    return config_dict, checksum
+
+
+def setup(filename: str) -> Dict:
+    """
+    Validate & apply configuration from disk
+
+    Args:
+        filename (str): path and filename to load.
+
+    Returns:
+        configuration data located at filename.
+    """
+    # do the loading part
+    config_dict, file_hash = load_config(filename)
+
+    # this is totally NOT a constant pylint. -_-
+    # also it appears the usage of global is unavoidable here, without introducing
+    # needless complexity via a class
+    global _hash  # pylint: disable=invalid-name, global-statement
+
+    if file_hash == _hash:
+        raise FileExistsError(f"new file's hash matches what we already have!"
+                              f"\told = {_hash}\tnew={file_hash}")
+
+    _hash = file_hash
+
+    setup_logging(config_dict['logging']['log_file'])
+    logging.info(f"new config hash is {file_hash}")
```

Comment:
[Suggestion] Demote to debug level logging?

Reply:
Im thinking it would be best to emit the fact a rehash occured, as well as the destination hash, as part of normal operation.

Perhaps the more verbose elements (such as the individual components handling things) would be debug level details.

Reply:
I agree that rehashes should be logged, but I just think that outputting the actual SHA256 is a bit overkill for normal operation.

Reply:
No kill quite like the overkill.

I would rather know the exact SHA256 of what it applied without bouncing mecha into verbose mode.